commit 08b36c4a02b5a9db609fc6d93b9c41d7fa75713c
Author: Paul Mundt <lethal@linux-sh.org>
Date:   Wed Jan 27 21:56:57 2010 +0900

    sh: Optimize runtime disabling of trapped I/O.
    
    Presently trapped I/O is only registered if it's not explicitly disabled
    for the platforms that select it openly. From the fault path this runs
    through an address lookup before figuring out that nothing matches and
    falls back through the error path, but we can forego the lookup
    completely by testing if it's been explicitly disabled. This provides a
    measurable speedup for things like qemu that rely on runtime disabling.
    
    Signed-off-by: Paul Mundt <lethal@linux-sh.org>

diff --git a/arch/sh/kernel/io_trapped.c b/arch/sh/kernel/io_trapped.c
index 16b83cc..4a8bb4e 100644
--- a/arch/sh/kernel/io_trapped.c
+++ b/arch/sh/kernel/io_trapped.c
@@ -271,6 +271,8 @@ int handle_trapped_io(struct pt_regs *regs, unsigned long address)
 	insn_size_t instruction;
 	int tmp;
 
+	if (trapped_io_disable)
+		return 0;
 	if (!lookup_tiop(address))
 		return 0;
 
